core/services/llo/observation/data_source.go

```diff
@@ -85,139 +100,285 @@ func NewDataSource(lggr logger.Logger, registry Registry, t Telemeter) llo.DataS
-			}
-		}(streamID)
-	}
+		// Notify the caller that we've completed our first round of observations.
+		if !d.observationLoopStarted.Load() {
+			d.observationLoopStarted.Store(true)
+			close(loopStartedCh)
+		}
 
-	// Wait for all Observations to complete
-	wg.Wait()
+		// After all Observations have returned, nothing else will be sent to the
+		// telemetry channel, so it can safely be closed
+		if telemCh != nil {
+			close(telemCh)
+		}
 
-	// Only log on errors or if VerboseLogging is turned on
-	if len(errs) > 0 || opts.VerboseLogging() {
-		elapsed := time.Since(now)
+		// Only log on errors or if VerboseLogging is turned on
+		if len(errs) > 0 || opts.VerboseLogging() {
+			slices.Sort(successfulStreamIDs)
+			sort.Slice(errs, func(i, j int) bool { return errs[i].streamID < errs[j].streamID })
 
-		slices.Sort(successfulStreamIDs)
-		sort.Slice(errs, func(i, j int) bool { return errs[i].streamID < errs[j].streamID })
+			failedStreamIDs := make([]streams.StreamID, len(errs))
+			errStrs := make([]string, len(errs))
+			for i, e := range errs {
+				errStrs[i] = e.String()
+				failedStreamIDs[i] = e.streamID
+			}
 
-		failedStreamIDs := make([]streams.StreamID, len(errs))
-		errStrs := make([]string, len(errs))
-		for i, e := range errs {
-			errStrs[i] = e.String()
-			failedStreamIDs[i] = e.streamID
-		}
+			lggr = logger.With(lggr, "elapsed", elapsed, "nSuccessfulStreams",
+				len(successfulStreamIDs), "nFailedStreams", len(failedStreamIDs), "errs", errStrs)
 
-		lggr = logger.With(lggr, "elapsed", elapsed, "nSuccessfulStreams",
-			len(successfulStreamIDs), "nFailedStreams", len(failedStreamIDs), "errs", errStrs)
+			if opts.VerboseLogging() {
+				lggr = logger.With(lggr, "streamValues", streamValues)
+			}
 
-		if opts.VerboseLogging() {
-			lggr = logger.With(lggr, "streamValues", streamValues)
+			if len(errs) == 0 && opts.VerboseLogging() {
+				lggr.Infow("Observation succeeded for all streamsToObserve")
+			} else if len(errs) > 0 {
+				lggr.Warnw("Observation failed for streamsToObserve")
+			}
 		}
 
-		if len(errs) == 0 && opts.VerboseLogging() {
-			lggr.Infow("Observation succeeded for all streams")
-		} else if len(errs) > 0 {
-			lggr.Warnw("Observation failed for streams")
+		promObservationLoopDuration.WithLabelValues(
+			opts.ConfigDigest().String()).Observe(float64(elapsed.Milliseconds()))
+
+		if elapsed < observationInterval {
+			lggr.Debugw("Observation loop sleep", "elapsed_ms", elapsed.Milliseconds(),
+				"interval_ms", observationInterval.Milliseconds(), "sleep_ms", observationInterval-elapsed)
+			time.Sleep(observationInterval - elapsed)
+		} else {
+			lggr.Debugw("Observation loop", "elapsed_ms", elapsed.Milliseconds(), "interval_ms", observationInterval.Milliseconds())
 		}
+
+		// Cancel the context, so the linter doesn't complain.
+		cancel()
+	}
+}
+
+func (d *dataSource) Close() error {
+	if !d.observationLoopStarted.Load() {
+		return nil
 	}
 
+	closeIfOpen(d.observationLoopCloseCh)
```

Comment:
This looks racey. Why is there any ambiguity about whether the channel needs to be closed?

Reply:
It appears to be set only from the constructor, so why not just:
```suggestion
	close(d.observationLoopCloseCh)
```

Reply:
There isn't any ambiguity at this point but there used to be in the past - we used to close the loop on critical errors and rely on the `Observe()` method to restart it.

I clean this up.